internal/host/plugin/repository_host_catalog.go

```diff
@@ -143,6 +146,270 @@ func (r *Repository) CreateCatalog(ctx context.Context, c *HostCatalog, _ ...Opt
+	}
+
+	// We also need the secrets for the host catalog - while we don't
+	// patch them in the same way as we do attributes, we do need them
+	// for the plugin request.
+	currentCatalogPersisted, err := r.getPersistedDataForCatalog(ctx, currentCatalog)
+	if err != nil {
+		return nil, db.NoRowsAffected, db.NoRowsAffected, errors.Wrap(ctx, err, op, errors.WithMsg(fmt.Sprintf("error looking up persisted data for catalog with id %q", c.PublicId)))
+	}
+
+	// Clone the catalog so that we can set fields.
+	newCatalog := currentCatalog.clone()
+	var dbMask, nullFields []string
+	for _, f := range fieldMask {
+		switch {
+		case strings.EqualFold("name", f) && c.Name == "":
+			nullFields = append(nullFields, "name")
+			newCatalog.Name = c.Name
+		case strings.EqualFold("name", f) && c.Name != "":
+			dbMask = append(dbMask, "name")
+			newCatalog.Name = c.Name
+		case strings.EqualFold("description", f) && c.Description == "":
+			nullFields = append(nullFields, "description")
+			newCatalog.Description = c.Description
+		case strings.EqualFold("description", f) && c.Description != "":
+			dbMask = append(dbMask, "description")
+			newCatalog.Description = c.Description
+		case strings.EqualFold("attributes", f) && c.Attributes != nil:
+			// Attributes are patched from the JSON included in the mask to
+			// the attributes that exist in the record.
+			dbMask = append(dbMask, "attributes")
+			newCatalog.Attributes, err = patchstruct.PatchBytes(newCatalog.Attributes, c.Attributes)
+			if err != nil {
+				return nil, db.NoRowsAffected, db.NoRowsAffected, errors.Wrap(ctx, err, op, errors.WithMsg("error in catalog attribute JSON"))
+			}
+
+			// Patch the working set with the new attributes.
+			c.Attributes = newCatalog.Attributes
+		case strings.EqualFold("secrets", f) && c.Secrets != nil:
+			// While in a similar format, secrets are passed along
+			// wholesale (for the time being). Don't append this mask
+			// field, as secrets do not have a database entry. Clear the
+			// secrets out of the working set after.
+			newCatalog.Secrets = c.Secrets
+			c.Secrets = nil
+
+		default:
+			return nil, db.NoRowsAffected, db.NoRowsAffected, errors.New(ctx, errors.InvalidFieldMask, op, fmt.Sprintf("invalid field mask: %s", f))
+		}
+	}
+
+	// Get the plugin client.
+	plgClient, ok := r.plugins[currentCatalog.GetPluginId()]
+	if !ok || plgClient == nil {
+		return nil, db.NoRowsAffected, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, fmt.Sprintf("plugin %q not available", currentCatalog.GetPluginId()))
+	}
+
+	// Convert the catalog values to API protobuf values, which is what
+	// we use for the plugin hook calls.
+	currPlgHc, err := toPluginCatalog(ctx, currentCatalog)
+	if err != nil {
+		return nil, db.NoRowsAffected, db.NoRowsAffected, errors.Wrap(ctx, err, op)
+	}
+	newPlgHc, err := toPluginCatalog(ctx, newCatalog)
```

Comment:
After reviewing this and trying out solutions that needed to be reverted due to lack of functionality in gorm and our own standards with regards to where such abstractions should lie, I'd recommend abandoning this for now. Here's my reasons:

Performing this work "around the edges" currently does not really give much value other than saving a single marshal/unmarshal cycle if attributes are to be updated. The values need to be unmarshaled anyway to be sent to the plugin, and as explained in Todd's message, service handlers will be doing more marshaling/unmarshaling anyway. This seems to need broad architectural changes to the data to effectively change this, probably out of the scope of this PR.
I still don't believe update would be in any hot path where performance is particularly important.
Concerns have been raised internally about storing a raw *structpb.Struct in the database. While I think we need to talk more about what can actually be done about it, given we don't know exactly the shape of attribute data looks like for plugin-based host catalogs, it might be a good idea to treat this data as opaquely as possible within the repository for now, and perform marshaling/unmarshaling as we need to do it.

For now, I'm just going to keep it as-is, move the PR to ready for review, and work on resolving any other comments.